File: 2150-kth-smallest-product-of-two-sorted-arrays/kth-smallest-product-of-two-sorted-arrays.ts
function f(nums2: number[], x1: number, v: number): number {
    const n2 = nums2.length;
    let left = 0,
        right = n2 - 1;
    while (left <= right) {
        const mid = Math.floor((left + right) / 2);
        const prod = nums2[mid] * x1;
        if ((x1 >= 0 && prod <= v) || (x1 < 0 && prod > v)) {
            left = mid + 1;
        } else {
            right = mid - 1;
        }
    }
    if (x1 >= 0) {
        return left;
    } else {
        return n2 - left;
    }
}

function kthSmallestProduct(
    nums1: number[],
    nums2: number[],
    k: number,
): number {
    const n1 = nums1.length;
    let left = -1e10,
        right = 1e10;
    while (left <= right) {
        const mid = Math.floor((left + right) / 2);
        let count = 0;
        for (let i = 0; i < n1; i++) {
            count += f(nums2, nums1[i], mid);
        }
        if (count < k) {
            left = mid + 1;
        } else {
            right = mid - 1;
        }
    }
    return left;
}

File: 2150-kth-smallest-product-of-two-sorted-arrays/kth-smallest-product-of-two-sorted-arrays.py
class Solution:
    def kthSmallestProduct(self, nums1: List[int], nums2: List[int], k: int) -> int:
        def calc_less_count(target):
            less_count = 0

            for num in nums1:
                if num == 0:
                    if target >= 0:
                        less_count += len(nums2)
                    continue
                
                if num < 0:
                    less_count += len(nums2) - bisect_left(nums2, ceil(target / num))
                else:
                    less_count += bisect_right(nums2, target // num)
            return less_count

        MIN = -(100_000 * 100_000)
        MAX = (100_000 * 100_000)
        l, r = MIN, MAX

        while l <= r:
            mid = (l + r) // 2
            if calc_less_count(mid) >= k:
                r = mid - 1
            else:
                l = mid + 1
        return l